File: Advent of Code 2020/day10.py
a = list(map(int, open("inputs/day10.in")))
a.sort()
a = [0] + a + [a[-1] + 3]

d1 = 0
d3 = 0

for i in range(1, len(a)):
    if a[i] - a[i - 1] == 1:
        d1 += 1
    if a[i] - a[i - 1] == 3:
        d3 += 1

print("Part 1:", d1 * d3)

options = [1] * len(a)

for i in range(len(a) - 2, -1, -1):
    options[i] = options[i + 1]
    if i + 2 < len(a) and a[i + 2] - a[i] < 4:
        options[i] += options[i + 2]
    if i + 3 < len(a) and a[i + 3] - a[i] < 4:
        options[i] += options[i + 3]

print("Part 2:", options[0])

File: Advent of Code 2021/lib/day10.ex
defmodule Day10 do
  def points_or_stack(line, stack \\ [], return_stack \\ false) do
    if String.length(line) == 0 do
      if return_stack do
        stack
        |> Enum.join()
      else
        0
      end
    else
      remainder = line |> String.slice(1..-1)

      case String.at(line, 0) do
        "(" ->
          Day10.points_or_stack(remainder, [")"] ++ stack, return_stack)

        "[" ->
          Day10.points_or_stack(remainder, ["]"] ++ stack, return_stack)

        "{" ->
          Day10.points_or_stack(remainder, ["}"] ++ stack, return_stack)

        "<" ->
          Day10.points_or_stack(remainder, [">"] ++ stack, return_stack)

        other ->
          if Enum.at(stack, 0) == other do
            Day10.points_or_stack(remainder, Enum.slice(stack, 1..-1), return_stack)
          else
            Map.get(%{")" => 3, "]" => 57, "}" => 1197, ">" => 25137}, other)
          end
      end
    end
  end

  def remainder_score(remainder, score \\ 0) do
    if String.length(remainder) == 0 do
      score
    else
      new_score =
        5 * score + Map.get(%{")" => 1, "]" => 2, "}" => 3, ">" => 4}, String.at(remainder, 0))

      remainder_score(String.slice(remainder, 1..-1), new_score)
    end
  end
end

{:ok, file} = File.read("inputs/day10.in")

lines =
  file
  |> String.split("\n", trim: true)

lines
|> Enum.map(&Day10.points_or_stack/1)
|> Enum.sum()
|> IO.inspect(label: "Part 1")

remainder_scores =
  lines
  |> Enum.filter(fn x -> Day10.points_or_stack(x) == 0 end)
  |> Enum.map(fn x -> Day10.points_or_stack(x, [], true) end)
  |> Enum.map(&Day10.remainder_score/1)
  |> Enum.sort()

remainder_scores
|> Enum.at(div(Enum.count(remainder_scores), 2))
|> IO.inspect(label: "Part 2")
